### ISSUE TITLE
Unable to perform asynchronous Node requests in "number of issues closed per contributors".

### ISSUE THREAD

**Issue by chessmasterhong** [DrkSephy/git-technetium#69]
Because of the way GitHub API's endpoints are structured, we need to perform several additional API requests to parse and build to create the full dataset. In the module "number of issues closed per contributors", we have the current procedure:
1. Get all contributor's names (1 request)
2. Initialize empty contributor data
3. Get all repository issues (1 request)
4. **For each** issue, get all issue events (1 request **per** issue)
5. Build the full dataset
6. Fill contributor data

It is at step 4 where the problem arises. Basically, there is a race condition with the async requests where a second request would get fired before the first request completes. It would yield the error "Can't set headers after they are sent." and would crash the Express server. However, we must include the header in our requests (specifically the user-agent) since the API would return with a 403 Forbidden error if one is not found ([https://developer.github.com/v3/#user-agent-required](https://developer.github.com/v3/#user-agent-required)).

**Comment by DrkSephy** [comment]
Could you post your asynchronous attempt at this, along with comments in the code pointing out where you got stuck? It would be helpful for everyone to see what you've tried and provides a good starting point for finding a solution.

**Comment by chessmasterhong** [comment]
I do not have the full code of the async attempt available anymore but it was probably something like this _(untested)_:

```
router.get('/issues_closed', function(req, res) {
        request({
            url: 'https://api.github.com/repos/' + req.query.ownerName + '/' + req.query.repoName + '/contributors',
            headers: { 'user-agent': 'git-technetium' },
            json: true
        }, function(error, response, body) {
            if(!error && response.statusCode === 200) {
                var contributors = [];
                for(var contributorIndex = 0; contributorIndex < body.length; contributorIndex++){
                    contributors.push(body[contributorIndex].login);
                }

                var contributorIssuesClosed = [];
                for(var contributorIndex = 0; contributorIndex < contributors.length; contributorIndex++) {
                    contributorIssuesClosed.push({
                        'name': contributors[contributorIndex],
                        'issues_closed': 0
                    });
                }

                request({
                    url: 'https://api.github.com/repos/' + req.query.ownerName + '/' + req.query.repoName + '/issues?state=closed',
                    headers: { 'user-agent': 'git-technetium' },
                    json: true
                }, function(error, response, body) {
                    if(!error && response.statusCode === 200) {
                        // async attempt begins here...

                        // initialize an empty queue of requests
                        var requestQueue = [];

                        // for each issue, push a request into the queue with each request being the events for the issue
                        for(var issueIndex = 0; issueIndex < body.length; issueIndex++) {
                            if(!body[issueIndex].pull_request) {
                                requestQueue.push(
                                    request({
                                        url: 'https://api.github.com/repos/' + req.query.ownerName + '/' + req.query.repoName + '/issues/' + (body[issueIndex].number - 1) + '/events',
                                        headers: { 'user-agent': 'git-technetium' },
                                        json: true
                                    }, function(error, response, body) {
                                        if(!error && response.statusCode === 200) {
                                            for(var eventIndex = body.length - 1; eventIndex >= 0; eventIndex--) {
                                                if(body[eventIndex].event === 'closed') {
                                                    for(var contributorIndex = 0; contributorIndex < contributors.length; contributorIndex++) {
                                                        if(body[eventIndex].actor.login === contributorIssuesClosed[contributorIndex].name) {
                                                            contributorIssuesClosed[contributorIndex].issues_closed++;
                                                            break;
                                                        }
                                                    }
                                                }
                                            }
                                        }
                                    });
                                );

                                // execute the request queue
                                require('async').series(requestQueue);

                                // send the resulting data after the request queue completes
                                res.send(contributorIssuesClosed);
                            }
                        }
                    }
                });
            }
        });
    });
```

**Comment by DrkSephy** [comment]
I've been working on this for the day, and here is what I have so far: https://gist.github.com/DrkSephy/7d1bae3d691f89f45ce0

I've commented the parts of the code that should be relevant. You'll need to install `async.js` via `npm install async` to run the code. 

Right now, I can compute the issues closed data for a set of 30 issues (this algorithm does not handle pagination yet). The requests are done using async.each, and in your console you can see the data being processed correctly. The problem is getting the data out. The callbacks do not have access to the data, if you try to access it, it will return as zero for each contributor. 

I know that async.waterfall allows you to pass data from one function to the next. Maybe an `async.each` wrapped inside an `async.waterfall` would work....but at this point I would like to get some feedback to see if my solution has helped us move forward at all. 

NOTE: The data being returned might be inconsistent from what Kevin's version returns. Either way, let's use this as a proof of concept to solve the underlying problem first. 

Feedback is much appreciated.

**Comment by DrkSephy** [comment]
This issue has been resolved by commit `b65df7245b64250b61717e63768cf9f72b59add2` and `7cd3ea342141df51a278b8bf7bfe287307a11e55`.
